### PR TITLE
Handle dynamic font color change

### LINKED ISSUE #18371: `color` attribute changes on the `<font>` element aren't respected.
## Testcase

```html
<!doctype html>
<font color="red">I should become green on click</font>
<script>
document.onclick = function() {
  document.querySelector('font').setAttribute('color', 'green');
}
</script>
```

Servo doesn't honor the attribute change at all, and keeps the text red.

## Task

The task is basically adding the necessary infrastructure to properly handles dynamic changes to mapped attributes.

Mapped attributes are attributes that change the style of a given element (like the `color` attribute of the `<font>` HTML element).

For reference (no need to understand that code to fix this bug), those attributes are passed to layout using the [`synthesize_presentational_hints_for_legacy_attributes`](https://github.com/servo/servo/blob/c60dd53210745d9d8e7d3a5ca0310370a33553f4/components/script/dom/element.rs#L473) API.

That function checks for a number of attributes, etc. However, the only code that handles invalidation of these kind of attributes lives in [`Document::element_attr_will_change`](https://github.com/servo/servo/blob/c60dd53210745d9d8e7d3a5ca0310370a33553f4/components/script/dom/document.rs#L2504), and right now it's a sad `if` that only handle `width` and `height` for any given element.

So, the task is adding the infrastructure to handle those dynamic changes easily.

In particular, we need to add a new function to the [`VirtualMethods`](https://github.com/servo/servo/blob/8b41c7c137c78aab208fe2991c3ec951fe5f1386/components/script/dom/virtualmethods.rs#L59) trait, that a given element will implement.

For now, we'll focus on keeping the existing behavior and fixing the `<font color>` bug, and adding an automated test for the last one.

### Adding `VirtualMethods::attribute_is_mapped`

We need to add an `attribute_is_mapped` function that returns whether a given attribute affects style or not for a given element.

The function should look something like:

```rust
fn attribute_is_mapped(&self, attr: &Attr) -> bool;
```

And should have a similar default implementation to the rest of the `VirtualMethods` types.

After that, we need to implement that function for the base `Element` "class". This function would do what the current code does now:

```rust
fn attribute_is_mapped(&self, attr: &Attr) -> bool {
    // FIXME: This should be more fine-grained, not all elements care about these.
    if attr.local_name() == &local_name!("width") ||
        attr.local_name() == &local_name!("height") {
        return true;
    }

    self.super_type().unwrap().attribute_is_mapped(attr)
}
```

After this, you need to override this function for `HTMLFontElement`, so that it looks like:

```rust
fn attribute_is_mapped(&self, attr: &Attr) -> bool {
    if attr.local_name() == &local_name!("color") {
        return true;
    }

    // FIXME: Should also return true for `size` and `face` changes!
    self.super_type().unwrap().attribute_is_mapped(attr)
}
```

(You're free to choose to fix also the `size` and `face` bugs too, but I'd recommend doing it as a followup or leaving it as an easy bug for somebody, because you'd need to write more tests and it'd take more to review and land).

## Calling the new method

Now we have all our setup in place, but unless you've gone ahead and modified the callsite in `document.rs` yourself, you'll see that the test-case is still broken.

We need to call our new virtual method instead of that `if` pointed at in the beginning!

To do so, we'd need to change that for something like:

```rust
if vtable_for(el.upcast()).is_attribute_mapped(attr) {
    entry.hint.insert(RESTYLE_SELF);
}
```

With that, hopefully the test-case should become green when you test and you can write an automated test for this!

We have documentation for that (see `CONTRIBUTING.md`), but please ask any question you may have!

Once you verify that your test passes, you're ready to submit a PR!

### PR DESCRIPTION
**Pull request by pylbrecht**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #18371.

- [X] There are tests for these changes

r? @emilio

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18405)

### CODE REVIEW

File: tests/wpt/mozilla/tests/css/font_color_attribute_change.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title></title>
+    <link rel="match" href="font_color_attribute_change_ref.html">
+  </head>
+  <body>
+    <font color="red">I should be green</font>
+    <script>
+      document.querySelector("font").setAttribute("color", "green");
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/font_color_attribute_change.html]: Let's add `document.body.offsetTop;` before this line, to ensure that layout has happened at least once.

**Review by emilio — COMMENTED** [review]
Looks good! Though I think you forgot to keep returning `true` for `width` and `height`, so I suspect this will fail a lot of tests.

**Comment by pylbrecht** [comment]
True, I forgot to check that file in.. I'll fix that and push again.

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -2303,6 +2303,16 @@ impl VirtualMethods for Element {
         Some(self.upcast::<Node>() as &VirtualMethods)
     }
 
+    fn attribute_is_mapped(&self, attr: &Attr) -> bool {
+        // FIXME: This should be more fine-grained, not all elements care about these.
+        if attr.local_name() == &local_name!("width") ||
+            attr.local_name() == &local_name!("height") {
+                return true;
```
Comment (emilio) [inline components/script/dom/element.rs]: nit: Indentation is off here.

Comment (pylbrecht) [reply components/script/dom/element.rs]: Thanks, vim indents it the wrong way for whatever reason.. fixed and pushed!

**Comment by emilio** [comment]
@bors-servo r+

Looks great, thanks!

Please ensure `./mach update-manifest` doesn't generate any diff, because otherwise CI will fail. If it does, please re-push, and ping me to get it merged.

Thanks a lot!

**Pull request merged**